packages/sync/src/modules/class-full-sync-immediately.php

```diff
@@ -0,0 +1,405 @@
+<?php
+/**
+ * Full sync module.
+ *
+ * @package automattic/jetpack-sync
+ */
+
+namespace Automattic\Jetpack\Sync\Modules;
+
+use Automattic\Jetpack\Sync\Defaults;
+use Automattic\Jetpack\Sync\Lock;
+use Automattic\Jetpack\Sync\Modules;
+use Automattic\Jetpack\Sync\Settings;
+
+/**
+ * This class does a full resync of the database by
+ * enqueuing an outbound action for every single object
+ * that we care about.
+ *
+ * This class, and its related class Jetpack_Sync_Module, contain a few non-obvious optimisations that should be explained:
+ * - we fire an action called jetpack_full_sync_start so that WPCOM can erase the contents of the cached database
+ * - for each object type, we page through the object IDs and enqueue them by firing some monitored actions
+ * - we load the full objects for those IDs in chunks of Jetpack_Sync_Module::ARRAY_CHUNK_SIZE (to reduce the number of MySQL calls)
+ * - we fire a trigger for the entire array which the Automattic\Jetpack\Sync\Listener then serializes and queues.
+ */
+class Full_Sync_Immediately extends Module {
+	/**
+	 * Prefix of the full sync status option name.
+	 *
+	 * @var string
+	 */
+	const STATUS_OPTION = 'jetpack_sync_full_status';
+
+	/**
+	 * Sync Lock name.
+	 *
+	 * @var string
+	 */
+	const LOCK_NAME = 'full_sync';
+
+	/**
+	 * Sync module name.
+	 *
+	 * @access public
+	 *
+	 * @return string
+	 */
+	public function name() {
+		return 'full-sync';
```

Comment:
Yes, the idea is that once we test proof this new "immediate" Full Sync, we will make it the default and only one. Less prone to errors since it doesn't have the entire Queue layer to fail.